homeassistant/helpers/translation.py

```diff
@@ -0,0 +1,129 @@
+"""Translation string lookup helpers."""
+import logging
+# pylint: disable=unused-import
+from typing import Optional  # NOQA
+from os import path
+
+from homeassistant.loader import get_component, bind_hass
+from homeassistant.util.json import load_json
+
+_LOGGER = logging.getLogger(__name__)
+
+TRANSLATION_STRING_CACHE = 'translation_string_cache'
+
+
+def recursive_flatten(prefix, data):
+    """Return a flattened representation of dict data."""
+    output = {}
+    for key, value in data.items():
+        if isinstance(value, dict):
+            output.update(
+                recursive_flatten('{}{}.'.format(prefix, key), value))
+        else:
+            output['{}{}'.format(prefix, key)] = value
+    return output
+
+
+def flatten(data):
+    """Return a flattened representation of dict data."""
+    return recursive_flatten('', data)
+
+
+def component_translation_file(component, language):
+    """Return the translation json file location for a component."""
+    if '.' in component:
+        name = component.split('.', 1)[1]
+    else:
+        name = component
+
+    module = get_component(component)
+    component_path = path.dirname(module.__file__)
+
+    # If loading translations for the package root, (__init__.py), the
+    # prefix should be skipped.
+    if module.__name__ == module.__package__:
+        filename = '{}.json'.format(language)
+    else:
+        filename = '{}.{}.json'.format(name, language)
+
+    return path.join(component_path, '.translations', filename)
+
+
+def load_translations_files(translation_files):
+    """Load and parse translation.json files."""
+    loaded = {}
+    for translation_file in translation_files:
+        loaded[translation_file] = load_json(translation_file)
+
+    return loaded
+
+
+def build_resources(translation_cache, components):
+    """Build the resources response for the given components."""
+    # Build response
+    resources = {}
+    for component in components:
+        if '.' not in component:
+            domain = component
+        else:
+            domain = component.split('.', 1)[0]
+
+        if domain not in resources:
+            resources[domain] = {}
+
+        # Add the translations for this component to the domain resources.
```

Comment:
Something I was thinking about the other day, especially now that we're doing more entity based config: should we store platform as an attribute in the state machine?

Reply:
The platform as an attribute would certainly be a big help here. It would let us completely avoid this problem and keep the translations namespaced separately. That's something we could add later though if we decide to make it available, as the conflict only exists in the API response. All of our translations at rest are stored separately per platform.

And FWIW this will only return translations for loaded components, so although we'll have lots of translations merged, each instance should only serve the translations it's using.